### PR TITLE
Adapt WordDetectorNN according to ADRs (closes #96)

### LINKED ISSUE #96: Adapt WordDetectorNN according to ADRs
Let's look at all the existing ADRs, in particular [ADR 006](https://github.com/PellelNitram/xournalpp_htr/blob/master/docs/ADRs/006_model_registry_and_training_environment.md), and let's implement the [WordDetectorNN](https://github.com/PellelNitram/xournalpp_htr/tree/master/xournalpp_htr/training/WordDetectorNN) model accordingly.

### PR DESCRIPTION
**Pull request by PellelNitram**
Closes #96.

Adapts the standalone `WordDetectorNN` subfolder to the architecture decided in the ADRs — in particular [ADR 006](docs/ADRs/006_model_registry_and_training_environment.md), consistent with ADR 002/003/005, and introducing [ADR 007](docs/ADRs/007_model_demos_local_only.md).

> History is intentionally **not squashed**. Two items were added and then changed in response to review — they show as paired commits: the `compute_predictions` pipeline (`bb0a9e6` add → `49faa32` remove) and the demo (`ab19cd4` CLI → `2f103c5` local Gradio). The summary below describes the **net result**.

## ADR 006 §1 — HF Hub as plain artifact storage
- New `xournalpp_htr/inference_models.py`: the verbatim `HFHubInferenceModel` ABC plus concrete `WordDetectorModel` (`HF_REPO_ID = PellelNitram/xournalpp-htr-word-detector`, parameter-free `from_pretrained()` via `hf_hub_download` + `onnxruntime`, model-specific `detect()`).

## ADR 006 §2 — ONNX as inference export format
- `word_detector/export.py` traces the checkpoint with softmax baked in to `model.onnx` + `config.json`. Uses the stable legacy exporter (control-flow-free CNN; avoids the `onnxscript` dep). `--upload` to HF Hub is documented and **not** run automatically.

## ADR 006 §3 — per-model extras & subfolders
- `pyproject.toml`: `scikit-learn`/`onnxruntime` added to base deps; new `training-word-detector` extra (`torch`, `torchvision`, `tensorboard`, `onnx`, `gradio`); no umbrella extra. `torch`/`torchvision` moved out of base deps into the training extra with a CUDA 12.8 index — inference needs only `onnxruntime`, no PyTorch.
- `training/WordDetectorNN/` (standalone `uv` project) → integrated `training/word_detector/`, monolithic `my_code.py` split into `network/dataset/train/export/infer/demo/utils`, with a dependency-guarded `__init__.py`.
- Base-deps-only shared utilities (`BoundingBox`, map decode/cluster/metrics) in `training/shared/`, importable in the lean inference install.

## ADR 006 §4 — repo naming
- `PellelNitram/xournalpp-htr-word-detector`.

## ADR 003 — central inference API (detection-only, no pipeline)
- WordDetector produces word **bounding boxes but no transcription/labels**. ADR 003's `compute_predictions` contract is word-level boxes *and* transcriptions, so WordDetector is **deliberately not exposed as a pipeline**. `models.py` is unchanged vs `master`. `WordDetectorModel` is the integration point; wiring into a full pipeline (incl. the ADR 005 document-unit conversion) is deferred until a recognition/labelling model exists. The earlier pipeline attempt was removed per review (`49faa32`).

## ADR 007 — local Gradio demo, no per-model HF Space
- New ADR 007. `word_detector/demo.py` is an interactive Gradio app launched **locally** (`demo.launch()`, optional `--share`); the HuggingFace Space, Supabase data-donation/telemetry, HF Space front-matter/`.env`, and the demo-architecture diagram are removed. Convention: every contributed model ships a local demo, no per-model Space.

## HF dataset integration
- New `load_IAM_DB_dataset()` in `xio.py` (mirrors `load_IAM_OnDB_dataset()`): resolves the IAM-DB training data via `snapshot_download` from `PellelNitram/xournalpp_htr_IAM_DB`.
- `train.py --data_path` defaults to `None`; when omitted, resolves via HF loader. Skips HF resolution entirely when the dataset pickle cache already exists (avoids slow HF validation on repeated runs).

## GPU training support
- `torch`/`torchvision` moved from base deps to `training-word-detector` extra with `pytorch-cu128` index — `uv sync --extra training-word-detector` installs CUDA-enabled PyTorch out of the box.
- Verified end-to-end on NVIDIA A100: training, export, demo, notebook all functional.

## Training sweep & model upload
- Hyperparameter grid search (lr × bs) via `run_training.sh`; best model: **lr=0.001, bs=16, F1=0.88**.
- `run_training.eval.sh` added to find the best model across sweep runs.
- Best model exported to ONNX and uploaded to `PellelNitram/xournalpp-htr-word-detector` with a model card.

## Notebook
- `word_detector/test_best_model.ipynb` (moved out of `notebooks/` subdirectory): exports checkpoint to ONNX, builds local `WordDetectorModel`, runs detection on example images with **PyTorch vs ONNX side-by-side comparison** (rescaled to original image coordinates), showing bounding box counts.

## Training runner
- `run_training.sh` fixed: invokes the renamed train module, `set -euo pipefail`, per-run `tee` logging, shebang, relative `BASE_PATH`.

## Tests / validation
- Offline tests: ABC contract, pure post-processing, and an ONNX round-trip (`slow`, skips without the local checkpoint / training extra).
- `make tests-not-slow`: 46 passed, 1 pre-existing failure (`test_run_htr` requires `xournalpp` binary), no regressions.

## Notes
- The migrated `tests.ipynb` (`legacy_tests.ipynb`) was removed: it imported the now-deleted `my_code` module and had broken top-level code. `test_best_model.ipynb` supersedes it.
- `from_pretrained()` now resolves — `model.onnx` + `config.json` have been uploaded to the model repo.
- Remaining TODOs (global step refactor, data augmentations) tracked in #102.

### CODE REVIEW

File: xournalpp_htr/training/word_detector/README.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+```python
+from xournalpp_htr.inference_models import WordDetectorModel
+
+model = WordDetectorModel.from_pretrained()           # or revision="v1.2.0"
+boxes = model.detect(grayscale_image)                 # list[BoundingBox]
+```
+
+It is also wired into the central inference API as the `word_detector_nn`
+pipeline (ADR 003); bounding boxes are returned in document units (ADR 005).
+WordDetector is detection-only and produces no transcription.
```
Comment (PellelNitram) [inline xournalpp_htr/training/word_detector/README.md]: This is a problem here because we don't want a `word_detector_nn` pipeline.

File: xournalpp_htr/training/word_detector/README.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+```python
+from xournalpp_htr.inference_models import WordDetectorModel
+
+model = WordDetectorModel.from_pretrained()           # or revision="v1.2.0"
+boxes = model.detect(grayscale_image)                 # list[BoundingBox]
+```
+
+It is also wired into the central inference API as the `word_detector_nn`
+pipeline (ADR 003); bounding boxes are returned in document units (ADR 005).
```
Comment (PellelNitram) [inline xournalpp_htr/training/word_detector/README.md]: This is a problem here because we don't want a word_detector_nn pipeline.

Comment (PellelNitram) [reply xournalpp_htr/training/word_detector/README.md]: That's because word_detector_nn is not a full pipeline but only produces bounding boxes but without labels

File: xournalpp_htr/models.py [hunk 3]
```diff
@@ -94,6 +94,51 @@ def compute_predictions(
                         )
                 predictions[page_index] = predictions_page
 
+    elif pipeline_name == "word_detector_nn":
```
Comment (PellelNitram) [inline xournalpp_htr/models.py]: let's remove all this because we don't need the pipeline

**Comment by PellelNitram** [comment]
ok, so i have reviewed all the 6 commits in the PR and here are some feedback notes that need fixing:

1. (also see above review) remove the word-detector pipeline because it does not do transcription. the model committed here is really just one model in a pipeline. the other model for transcription purposes will be added later
2. run the new code once i am on a GPU enabled machine. also, as such, run the tests added
3. automate the deployment of the gradio demo through CI/CD. alternatively, remove the online HF gradio demo in favour of only a local demo. also note that all future models that are contributed should feature a local demo just to check if the trained model works well or not. since also adding a HF demo for each new model feels a bit much work, i'd probably vote for removing also the HF demo for this model
4. Make sure to remove the mention of a pipeline in the model's readme once the pipeline is removed.
5. adapt code to use HF dataset instead of local dataset; i might need to upload the local dataset to HF for that (probably as a private one for license purposes)
6. adapt run_training.sh script to be functional (e.g. use HF dataset and avoid local paths) - also, maybe use a python script for running the experiments as it might make the running easier (?). on the other hand, a bash script is great for that and very pragmatic b/c it gets the job done (aka make it reproducible)
7. look into TODOs and maybe implement them
8. see future experiments in `run_training.sh`

**Comment by PellelNitram** [comment]
Addressed review: removed the `word_detector_nn` branch from `compute_predictions` (`models.py` is now identical to `master`) and rewrote the README inference section. Rationale captured: WordDetector is detection-only (boxes, no labels), so it is **not** exposed as an ADR-003 `compute_predictions` pipeline; `WordDetectorModel` remains the integration point until a recognition/labelling stage exists. Commit 49faa32.

**Comment by PellelNitram** [comment]
The 7th commit in this PR looks also good to me

**Comment by PellelNitram** [comment]
Addressed review item **#3 (demo)** and **#4 (write it down)**: dropped the HuggingFace Gradio Space + Supabase data-donation path in favour of a **local-only** demo (lower maintenance, no infra/secrets, runs offline). Recorded as a project-wide convention in new [ADR 007](docs/ADRs/007_model_demos_local_only.md): every contributed model ships a local demo; no per-model HF Space.

- `events.py` → `utils.py` (Supabase `save_event` removed; git-hash/JSON-encoder/example-list kept)
- `demo.py` rewritten as a no-UI CLI: runs a checkpoint on an image (or the bundled examples) and writes annotated output
- README: HF Space front-matter + Supabase/deployment sections removed; `Local demo` usage + ADR-007 convention added
- removed `.env.example` and the gradio-architecture excalidraw (artifacts of the dropped Space)

Commit ab19cd4. Lint clean; 16 offline tests green.

**Comment by PellelNitram** [comment]
Correction on the demo: kept the **interactive Gradio app**, just **not deployed as a HuggingFace Space** (no Supabase/telemetry, no hosted Space).

- `demo.py` is a local `gr.Interface` launched via `demo.launch()` — image + margin slider in, predicted boxes out. `--device {cpu,cuda,auto}`, optional `--share` for an ad-hoc public link. No data-donation / Supabase / HF Space front-matter.
- `gradio` moved into the `training-word-detector` extra so `uv sync --extra training-word-detector` covers training + export + the local demo.
- ADR 007 retitled/revised: *Local Gradio App, No Per-Model HuggingFace Space* (UI kept; only hosting + telemetry dropped).
- README demo section updated accordingly.

Commit 2f103c5 (supersedes ab19cd4's CLI approach). Lint clean; 16 offline tests green; demo imports and `--help` verified.

File: xournalpp_htr/training/word_detector/README.md (hunk 2 above)
Comment (PellelNitram) [reply xournalpp_htr/training/word_detector/README.md]: done :heavy_check_mark:

**Comment by PellelNitram** [comment]
Addressed review **#6a** (`run_training.sh`) — kept bash, fixed the bugs:

- `python run_training.py` → `uv run python -m xournalpp_htr.training.word_detector.train` (renamed module)
- added `#!/usr/bin/env bash` shebang
- `set -e` → `set -euo pipefail` so a failed run still aborts despite the `tee` pipe
- resolves the file's own logging TODO: per-run `mkdir -p` + `2>&1 | tee $OUT/train.log`
- removed the dangling trailing `\` (continued to nothing); quoted variables

Left as-is per your call: the `BASE_PATH`/data path (you'll do the HF-dataset path fix later, #5) and the *Future experiments* comment block. `bash -n` clean. Commit dbfc70b.

**Comment by PellelNitram** [comment]
Addressed review **#5 (HF dataset)** and the data-path part of **#6 (`run_training.sh`)** — commit cdb9b5d.

- New `load_IAM_DB_dataset()` in `xio.py`, a faithful twin of `load_IAM_OnDB_dataset()` (same `huggingface_hub` import guard; `snapshot_download(repo_id="PellelNitram/xournalpp_htr_IAM_DB", repo_type="dataset") / "data"`). The IAM-DB forms are raw files, so this resolves a cached path rather than using `load_dataset`.
- `IAM_Dataset` directory constants renamed to the HF layout: `gt` → `xml`, `img` → `forms`. The stem-pairing logic and extensions were already compatible, so this is the only structural delta.
- `train.py --data_path` default `→ None`; when omitted it resolves via the HF loader (printed + recorded in `args.json`). Removes the dead `~/Development/WordDetectorNN/data/train` default; `--data_path` still overrides for offline/custom copies.
- `run_training.sh` needs **no change**: it never passed `--data_path`, so the new HF-resolving default applies automatically. `BASE_PATH` stays as a legitimate user-local *output* dir, so no local *data* paths remain.

Convention follows the existing IAM-OnDB pattern: the `Dataset` class stays path-only; HF resolution lives in an `xio.py` `load_*` helper.

Verified offline against the already-cached `IAM_DB` snapshot: path resolves, `forms/` + `xml/` present, all **1539/1539** XML↔PNG pairs match, `_parse_gt` yields labelled boxes on a sample. Pre-commit `ruff`/`ruff-format` clean. Note: full `make tests-not-slow` was **not** re-run to completion (aborted to avoid an HF rate-limit clash with a concurrent download); rerun pending once the IAM-OnDB cache is warm.

Still deferred per earlier discussion: review **#2** (GPU run of code + added tests) and the *Future experiments* block (**#8**).
